File: src/problemset/dfs/KeysAndRooms.py
''' -*- coding: utf-8 -*-
@date: 2020-03-25
@time: 18:43:36
@author: MoeexT
Problem: 841.
url: https://leetcode-cn.com/problems/keys-and-rooms/
'''
from functools import reduce
from typing import List


class Solution:
    def __init__(self):
        self.flag = [];
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        '''
        @time: 104ms 31.73%
        @mem: 14.2MB 6.56%
        '''
        self.flag = [False for i in range(len(rooms))]
        self.flag[0] = True
        self.dfs(rooms, 0)
        print(self.flag)
        return reduce(lambda x, y: x and y, self.flag)
        
    def dfs(self, rooms, idx):
        for key in rooms[idx]:
            if self.flag[key]:
                continue
            self.flag[key] = True
            self.dfs(rooms, key)



if __name__ == "__main__":
    solution = Solution()
    print(solution.canVisitAllRooms([[]]))

File: src/problemset/dfs/KeysAndRooms.java
package problemset.dfs;

import java.util.ArrayList;
import java.util.Arrays;

/** 
 * date: 2020-03-25
 * time: 17:07:55
 * author: MoeexT
 * Problem: 841
 * url: https://leetcode-cn.com/problems/keys-and-rooms/
 * 传入`ArrayList<ArrayList<Integer>> rooms`参数不对
 */

public class KeysAndRooms {
    boolean[] flag;
    public boolean canVisitAllRooms (ArrayList<ArrayList<Integer>> rooms) {
        flag = new boolean[rooms.size()];
        boolean res = true;
        // Stream.of(flag).reduce(true, (a, b) -> a && b);

        dfs(rooms, 0);

        for (boolean b : flag) {
            res = res && b;
        }
        return res;
    }

    public void dfs(ArrayList<ArrayList<Integer>> rooms, int idx) {
        for (int key : rooms.get(idx)) {
            if (flag[key])
                continue;
            flag[key] = true;
            dfs(rooms, key);
        }
    }

    public static void main(String[] args) {
        KeysAndRooms obj = new KeysAndRooms();
        System.out.println(obj.canVisitAllRooms(new ArrayList<ArrayList<Integer>>(Arrays.asList(
            new ArrayList<Integer>(Arrays.asList(1)),
            new ArrayList<Integer>(Arrays.asList(2)),
            new ArrayList<Integer>(Arrays.asList(3)),
            new ArrayList<Integer>(Arrays.asList())
        ))));
    }
}
